tests/ets.rs

```diff
@@ -1,13 +1,110 @@
+
+use rand::Rng;
+use rusoto::{ChainProvider, ProvideAwsCredentials, Region};
 use rusoto::ets::EtsClient;
-use rusoto::{ChainProvider, Region};
+use rusoto::s3::{BucketName, S3Helper};
 
 static AWS_REGION: Region = Region::UsEast1;
+static AWS_SERVICE_RANDOM_SUFFIX_LENGTH: usize = 50;
+
+struct TestEtsClient<P>
+    where P: ProvideAwsCredentials
+{
+    credentials_provider: P,
+    region: Region,
+
+    client: EtsClient<P>,
+
+    s3_helper: Option<S3Helper<P>>,
+    input_bucket: Option<BucketName>,
+    output_bucket: Option<BucketName>,
+}
+
+impl<P> TestEtsClient<P>
+    where P: ProvideAwsCredentials + Clone
+{
+    /// Creates a new `EtsClient` for a test.
+    fn new(credentials_provider: P, region: Region) -> Self {
+        TestEtsClient {
+            credentials_provider: credentials_provider.clone(),
+            region: region,
+            client: EtsClient::new(credentials_provider, region),
+            s3_helper: None,
+            input_bucket: None,
+            output_bucket: None,
+        }
+    }
+
+    fn create_s3_helper(&mut self) {
+        self.s3_helper = Some(S3Helper::new(
+            self.credentials_provider.clone(),
+            self.region
+        ));
+    }
+
+    fn create_bucket(&mut self) -> BucketName {
+        let bucket_name = generate_unique_name("ets-bucket-1");
+        let result = self.s3_helper
+            .as_mut()
+            .unwrap()
+            .create_bucket_in_region(&bucket_name, AWS_REGION, None);
+        let mut location = result
+            .unwrap()
+            .location;
+        // A `Location` is identical to a `BucketName` except that it has a
+        // forward slash prepended to it, so we need to remove it.
+        location.remove(0);
+        info!("Created S3 bucket: {}", location);
+        location
+    }
+}
+
+impl<P> Deref for TestEtsClient<P>
+    where P: ProvideAwsCredentials
+{
+    type Target = EtsClient<P>;
+    fn deref(&self) -> &Self::Target {
+        &self.client
+    }
+}
+
+impl<P> DerefMut for TestEtsClient<P>
+    where P: ProvideAwsCredentials
+{
+    fn deref_mut<'a>(&'a mut self) -> &'a mut EtsClient<P> {
+        &mut self.client
+    }
+}
+
+impl<P> Drop for TestEtsClient<P>
+    where P: ProvideAwsCredentials
+{
+    fn drop(&mut self) {
+        self.s3_helper.take().map(|mut s3_helper| {
+            self.input_bucket.take().map(|bucket| {
+                match s3_helper.delete_bucket(&bucket, self.region) {
+                    Ok(_) => { info!("Deleted S3 bucket: {}", bucket) },
+                    Err(e) => { error!("Failed to delete S3 bucket: {}", e) }
+                };
+            });
+            self.output_bucket.take().map(|bucket| {
+                match s3_helper.delete_bucket(&bucket, self.region) {
+                    Ok(_) => { info!("Deleted S3 bucket: {}", bucket) },
+                    Err(e) => { error!("Failed to delete S3 bucket: {}", e) }
+                };
+            });
+        });
```

Comment:
I'll leave it as is since `if let` blocks wouldn't really make this shorter and I think that writing things functionally where possible is more "rustic".